File: geekbang/algorithm/source-code/2/lc49 字母异位词分组.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
/*
{
    "aet": ["eat", "tea", "ate"],
    "ant": ["tan", "nat"],
    "abt": ["bat"]
}
*/
        unordered_map<string, vector<string>> groups;
        for (string& s : strs) {
            string copy = s;
            sort(copy.begin(), copy.end());
            if (groups.find(copy) == groups.end()) {
                groups[copy] = {};
            }
            groups[copy].push_back(s);
        }
        vector<vector<string>> ans;
        for (const pair<string, vector<string>>& group : groups) {
            ans.push_back(group.second);
        }
        return ans;
    }
};

File: geekbang/algorithm/source-code/2/lc49 字母异位词分组.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        '''
        {
            "aet": ["eat", "tea", "ate"],
            "ant": ["tan", "nat"],
            "abt": ["bat"]
        }
        '''
        groups = {}
        for s in strs:
            copy = ''.join(sorted(s))
            if copy not in groups:
                groups[copy] = []
            groups[copy].append(s)
        ans = []
        for group in groups.values():
            ans.append(group)
        return ans